include/ada/parser.h

```diff
@@ -0,0 +1,29 @@
+#ifndef ADA_PARSER_H
+#define ADA_PARSER_H
+
+#include "state.h"
+#include "url.h"
+
+#include <optional>
+#include <string_view>
+
+namespace ada::parser {
+
+  // Return `std::nullopt` if the parser result is failure.
+  template<typename T> using output = std::optional<T>;
+  template<typename T> using parser_result = std::tuple<output<T>, bool>;
+
+  parser_result<std::string_view> domain_to_ascii(const std::string_view input, bool be_strict = false);
```

Comment:
@lemire What do you think about these types of definitions?

Reply:
Depending on the actual implementation, it might be beneficial to use a template... e.g.,
```
struct a{
    template <bool b=false>
    int f() {
        return b;
    }
};

int z() {
    a a;
    return a.f();
}
```

The implementation would be much the same. The compiler would, however, create two distinct functions. If you expect most people to rely on 'be_string=false', then it is possible that the commonly used function could be faster.

(Note that with C++11, you can now define template functions inside source files so it is really no different.)